src/sentry/backup/validate.py

```diff
@@ -21,60 +23,97 @@ def validate(
     list of reasons why not when it doesn't.
     """
 
+    class OrdinalCounter:
+        """Keeps track of the next ordinal to be assigned for a given model kind."""
+
+        max_seen_pk: int
+        next_ordinal: int
+
+        def __init__(self):
+            self.max_seen_pk = -1
+            self.next_ordinal = 1
+
+        def assign(self, obj: JSONData, side: Side) -> Tuple[int, list[ComparatorFinding]]:
+            """Assigns the next available ordinal to the supplied `obj` model."""
+
+            pk = obj["pk"]
+            model = obj["model"]
+            findings = []
+            if pk > self.max_seen_pk:
+                self.max_seen_pk = pk
+            else:
+                findings.append(
+                    ComparatorFinding(
+                        kind="UnorderedInput",
+                        on=InstanceID(model, self.next_ordinal),
+                        left_pk=pk if side == Side.left else None,
+                        right_pk=pk if side == Side.right else None,
+                        reason=f"""instances not listed in ascending `pk` order; `pk` {pk} is less than or equal to {self.max_seen_pk} which precedes it""",
+                    )
+                )
+
+            obj["ordinal"] = self.next_ordinal
+            self.next_ordinal += 1
+            return (obj["ordinal"], findings if findings else [])
+
+    OrdinalCounters = Dict[str, OrdinalCounter]
+    ModelMap = Dict[InstanceID, JSONData]
+
+    def build_model_map(
+        models: JSONData, side: Side, findings: ComparatorFindings
+    ) -> Tuple[ModelMap, OrdinalCounters]:
+        """Does two things in tandem: builds a map of InstanceID -> JSON model, and simultaneously builds a map of model name -> number of ordinals assigned."""
+
+        model_map: ModelMap = {}
+        ordinal_counters: OrdinalCounters = {}
+        for model in models:
+            model_name = model["model"]
+            ordinal_counters.setdefault(model_name, OrdinalCounter())
```

Comment:
nit: might not need to do this is we use `defaultdict`

Reply:
Whoops, missed this comment. Will change in a follow up PR.

Reply:
Fixed by PR #54476